src/v/kafka/server/handlers/topics/topic_utils.cc

```diff
@@ -48,13 +50,17 @@ ss::future<> wait_for_leaders(
             // topic already deleted
             continue;
         }
+        /**
+         * Partition space is contiguous, use integer range to iterate over all
+         * topic partitions.
+         */
+        int32_t partition_count = md->get().get_assignments().size();
         co_await ss::max_concurrent_for_each(
-          md->get().get_assignments(),
+          std::views::iota(0, partition_count),
           64,
-          [&r, &md_cache, deadline](const auto& p_as) {
+          [&r, &md_cache, deadline](int32_t p_id) {
```

Comment:
> When iterating ... it may happen the assignment is removed.

Clearly the loop is safe in the absence of the problematic concurrent modifications. But what is missing for me is knowing if the problem is that the concurrent modification exists (the fix is elsewhere), or if the loop was always expected to be tolerant of concurrent modification to the container.

Reply:
In other places that require the information we have a special type of iterator that validates if the concurrent modification happened. I would change this place if it would actually require the information from the partition assignment object. I realised that in this case the only piece that is required is the partition id. This is why i decided to use the simple integer range.

Reply:
>  if the problem is that the concurrent modification exists 

It seems like it will definitely exist? By concurrent we mean just some unrelated task. The modification can happen due to partition operations which remove partitions/topics. Since that can happen at "any time" and there are no locks taken, it seems like concurrent modifications can always occur for topic metadata. 

I think it's more or less true about any structure that can be modified and is shared within a shard: other than having explicit locks (which there are not for this metadata stuff), I mostly seems impossible to exclude concurrent modification?

Reply:
> It seems like it will definitely exist?

yeh fair